### PR TITLE
[FLINK-10755][table] Port external catalogs in Table API extension points to flink-table-common

### PR DESCRIPTION
**Pull request by dianfu**
## What is the purpose of the change

*This pull request ported external catalog to flink-table-common*

## Brief change log

  - *Ported the following classes to flink-table-common: ExternalCatalog, ExternalCatalogTable, CatalogNotExistException, TableNotExistException*
  - *Remove method ExternalCatalogTable.builder()*

## Verifying this change

This change is a code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by dianfu** [comment]
@twalthr Have created a PR to port external catalogs to flink-table-common. Could you help to take a look when it's convenient for you? Thanks in advance. :)

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ExternalCatalogTable.java [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+
+/**
+ * Defines a table in an {@link ExternalCatalog}. External catalog tables describe table sources
+ * and/or sinks for both batch and stream environments.
+ *
+ * <p>See also {@link TableFactory} for more information about how to target suitable factories.
+ *
+ * <p>Use {@link ExternalCatalogTableBuilder} to integrate with the normalized descriptor-based API.
+ */
+@PublicEvolving
```
Comment (bowenli86) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ExternalCatalogTable.java]: remove annotation

**Comment by dianfu** [comment]
@bowenli86 Thanks a lot for the suggestions. I'm fine to remove the annotation of 
ExternalCatalog and ExternalCatalogTable if they will be replaced with new Catalog API as these classes may be refactored or removed at that time. Regarding to the annotation itself, I think it's needed regardless whether it exists before. Porting classes is a good opportunity to correct the annotation which may be not marked correctly before. @twalthr What's your thought?

**Comment by sunjincheng121** [comment]
@flinkbot approve description
@flinkbot approve consensus

Should be rebased after all commits of FLINK-11449 be merged.

**Comment by dianfu** [comment]
@sunjincheng121 Thanks a lot for the review. I will rebase the PR after 
all the commits of FLINK-11449 merged. :)

**Comment by dianfu** [comment]
@sunjincheng121 @twalthr I have rebased the PR since FLINK-11449 has been merged. Could you help to take a look at this PR? Thanks in advance.

**Comment by twalthr** [comment]
@bowenli86 thanks for your suggestion. I agree that the catalog API will change in the near future but this doesn't mean that we can't annotate it. The annotation allows users to know what is intended to be used or not. The current catalog APIs are still intended to be used. The `Evolving` tells people that it might change again. Additionally, the annotation tells contributors when to think twice about changing a class.

**Comment by bowenli86** [comment]
@twalthr re: The current catalog APIs are still intended to be used. What my team is thinking of rolling out unified catalog APIs is to build them separately and not touch existing ExternalCatalog APIs, and, upon it having full capabilities of existing catalog APIs, we can just do a switch. That's how we did it internally, and of course we can discuss if you have a different rollout plan. Anyway, the above was the background I had in mind then. With that, I was concerned that japicmp may report errors when changing a class's annotation from PublicEvolving to Deprecated, thus I did some testing locally and it seems to be fine.

I think it's ok to mark them as PublicEvolving. It's a bit hard to say as everything is moving around so fast in Flink table/SQL and our planned timeline.

**Comment by twalthr** [comment]
@bowenli86 Having a completely new interface for the unified catalog APIs and performing the switch later is completely fine. This is also what Xuefu and I had discussed offline. As far as I know, `japicmp` is not in place for the Table & SQL API.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ExternalCatalogTable.java [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+	 * <p>It takes {@link Descriptor}s which allow for declaring the communication to external
+	 * systems in an implementation-agnostic way. The classpath is scanned for suitable table
+	 * factories that match the desired configuration.
+	 *
+	 * <p>Use the provided builder methods to configure the external catalog table accordingly.
+	 *
+	 * <p>The following example shows how to read from a connector using a JSON format and
+	 * declaring it as a table source:
+	 *
+	 * <p>{@code
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ExternalCatalogTable.java]: Use `<pre>{@code …}</pre>` instead. See https://reflectoring.io/howto-format-code-snippets-in-javadoc/.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/sources/TableSourceUtil.scala [hunk 3]
```diff
@@ -374,15 +374,17 @@ object TableSourceUtil {
...
       } else {
-        new Array[ResolvedFieldReference](0)
+        new Array[FieldReferenceExpression](0)
       }
 
       val expression = tsExtractor.getExpression(fieldAccesses)
       // add cast to requested type and convert expression to RexNode
-      // TODO we cast to planner expressions as a temporary solution to keep the old interfaces
-      val rexExpression = Cast(expression.asInstanceOf[PlannerExpression], resultType)
+      val rexExpression = ApiExpressionUtils
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/sources/TableSourceUtil.scala]: I would leave a TODO here as it reminds us that this is still not the final solution.

File: flink-connectors/flink-connector-elasticsearch6/pom.xml [hunk 4]
```diff
@@ -172,6 +172,14 @@ under the License.
 			<scope>test</scope>
 		</dependency>
 
+		<dependency>
+			<groupId>org.apache.flink</groupId>
```
Comment (twalthr) [inline flink-connectors/flink-connector-elasticsearch6/pom.xml]: Can we remove dependencies to `flink-table-api-java-bridge` or `flink-table-planner` already now? I think the elasticsearch module should be able to work with `flink-table-common` classes only, right?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Rowtime.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+		return this;
+	}
+
+	/**
+	 * Sets a custom timestamp extractor to be used for the rowtime attribute.
+	 *
+	 * @param extractor The {@link TimestampExtractor} to extract the rowtime attribute
+	 *                  from the physical type.
+	 */
+	public Rowtime timestampsFromExtractor(TimestampExtractor extractor) {
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Rowtime.java]: This changes the original logic. Before we were only serializing as the last option. Serializing is dangerous as we might not be able to restore the base64 string. Please reintroduce the original logic.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Rowtime.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+		internalProperties.putString(
+			ROWTIME_WATERMARKS_TYPE, ROWTIME_WATERMARKS_TYPE_VALUE_FROM_SOURCE);
+		return this;
+	}
+
+	/**
+	 * Sets a custom watermark strategy to be used for the rowtime attribute.
+	 */
+	public Rowtime watermarksFromStrategy(WatermarkStrategy strategy) {
+		internalProperties.putString(
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Rowtime.java]: This changes the original logic. Before we were only serializing as the last option. Serializing is dangerous as we might not be able to restore the base64 string. Please reintroduce the original logic.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Rowtime.java [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+/**
+ * Rowtime descriptor for describing an event time attribute in the schema.
+ */
+@PublicEvolving
+public class Rowtime implements Descriptor {
+
+	private final DescriptorProperties internalProperties = new DescriptorProperties(true);
+
+	/**
+	 * These constants will be removed once RowtimeValidator is ported to flink-table-common.
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Rowtime.java]: Is it too much effort to port the `RowtimeValidator` now?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Schema.java [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+		lastField = fieldName;
+		return this;
+	}
+
+	/**
+	 * Specifies the origin of the previously defined field. The origin field is defined by a
+	 * connector or format.
+	 *
+	 * <p>E.g. field("myString", Types.STRING).from("CSV_MY_STRING")
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Schema.java]: Use `{@code field(...}` for all code exmaples in those methods.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/sources/FieldComputer.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+	 */
+	public abstract void validateArgumentFields(TypeInformation<?>[] argumentFieldTypes) throws ValidationException;
+
+	/**
+	 * Returns the {@link Expression} that computes the value of the field.
+	 *
+	 * @param fieldAccesses Field access expressions for the argument fields.
+	 * @return The expression to extract the timestamp from the {@link TableSource} return type.
+	 */
+	public abstract Expression getExpression(FieldReferenceExpression[] fieldAccesses);
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/sources/FieldComputer.java]: I had this discussion with @sunjincheng121 before that this breaks the current interfaces and also changes the serialization format of serialized timestamp extractors. I'm not happy with this changes but I also see that we need progress here.

In any case, I would like to get such breaking changes communicated through the description of the PR.

If we would simply postpone the `ExternalCatalogBuilder` porting, most of the work of this PR would not have been necessary.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/descriptors/RowtimeTest.scala [hunk 10]
```diff
@@ -87,17 +87,17 @@ class RowtimeTest extends DescriptorTestBase {
         "B0b3JzLlJvd3RpbWVUZXN0JEN1c3RvbUFzc2lnbmVyeDcuDvfbu0kCAAB4cgBHb3JnLmFwYWNoZS5mbGluay" +
         "50YWJsZS5zb3VyY2VzLndtc3RyYXRlZ2llcy5QdW5jdHVhdGVkV2F0ZXJtYXJrQXNzaWduZXKBUc57oaWu9A" +
         "IAAHhyAD1vcmcuYXBhY2hlLmZsaW5rLnRhYmxlLnNvdXJjZXMud21zdHJhdGVnaWVzLldhdGVybWFya1N0cm" +
-        "F0ZWd5mB_uSxDZ8-MCAAB4cA")
+        "F0ZWd57RqMYVyVWhUCAAB4cA")
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/descriptors/RowtimeTest.scala]: We should not break tests in "porting" pull requests.

**Comment by dianfu** [comment]
@twalthr Thanks a lot for the review. Considering that this change will break the API of FieldComputer and the serialization of Rowtime, I'd like to take the original solution and avoid the changes of ExternalCatalogBuilder. What do you think? I intended to avoid removing the method ExternalCatalogTable.builder() before (as done in the current PR), but it seems that removing it would be a nicer solution.

**Comment by twalthr** [comment]
@dianfu thanks for your feedback. Yes, I think we should always aim for small and less invasive steps. The external catalog interfaces will be deprecated anyway soon. If we change the rowtime descriptor and field computer interfaces, we should do it as part of a big table source refactoring. The most important goal is to unblock `Table` and `TableEnvironment` refactoring tasks.

File: flink-connectors/flink-connector-elasticsearch6/pom.xml (hunk 4 above)
Comment (dianfu) [reply flink-connectors/flink-connector-elasticsearch6/pom.xml]: It still depends on `flink-table-api-java-bridge` and `flink-table-planner` as `Elasticsearch6UpsertTableSinkFactory` implements `StreamTableSinkFactory`.

**Comment by dianfu** [comment]
@twalthr Makes much sense to me. I have updated the PR and limit the changes to a minimal. Looking forward to your feedback. :)

**Pull request closed without merging**